app/scripts/metamask-controller.js

```diff
@@ -5322,27 +5326,207 @@ export default class MetamaskController extends EventEmitter {
   }
 
   /**
-   * Requests approval for permissions for the specified origin
+   * Prompts the user with permittedChains approval for given chainId.
    *
-   * @param origin - The origin to request approval for.
-   * @param permissions - The permissions to request approval for.
+   * @param {string} origin - The origin to request approval for.
+   * @param {Hex} chainId - The chainId to add incrementally.
    */
-  async requestPermissionApprovalForOrigin(origin, permissions) {
+  async requestApprovalPermittedChainsPermission(origin, chainId) {
     const id = nanoid();
-    return this.approvalController.addAndShowApprovalRequest({
+    await this.approvalController.addAndShowApprovalRequest({
       id,
       origin,
       requestData: {
         metadata: {
           id,
           origin,
         },
-        permissions,
+        permissions: {
+          [PermissionNames.permittedChains]: {
+            caveats: [
+              {
+                type: CaveatTypes.restrictNetworkSwitching,
+                value: [chainId],
+              },
+            ],
+          },
+        },
       },
       type: MethodNames.RequestPermissions,
     });
   }
 
+  /**
+   * Requests permittedChains permission for the specified origin
+   * and replaces any existing CAIP-25 permission with a new one.
+   *
+   * @param {string} origin - The origin to request approval for.
+   * @param {Hex} chainId - The chainId to add incrementally.
+   * @param {boolean} autoApprove - If the chain should be granted without prompting for user approval.
+   */
+  async requestPermittedChainsPermission(origin, chainId, autoApprove) {
+    if (isSnapId(origin)) {
+      throw new Error(
+        `cannot request permittedChains permission for snaps with origin "${origin}"`,
+      );
+    }
+
+    if (!autoApprove) {
+      await this.requestApprovalPermittedChainsPermission(origin, chainId);
+    }
+
+    let caveatValue = {
+      requiredScopes: {},
+      optionalScopes: {},
+      isMultichainOrigin: false,
+    };
+    caveatValue = addPermittedEthChainId(caveatValue, chainId);
+
+    this.permissionController.grantPermissions({
+      subject: { origin },
+      approvedPermissions: {
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caveatValue,
+            },
+          ],
+        },
+      },
+    });
+  }
+
+  /**
+   * Requests incremental permittedChains permission for the specified origin.
+   * and updates the existing CAIP-25 permission.
+   *
+   * @param {string} origin - The origin to request approval for.
+   * @param {Hex} chainId - The chainId to add incrementally.
+   * @param {boolean} autoApprove - If the chain should be granted without prompting for user approval.
+   */
+  async requestPermittedChainsPermissionIncremental(
+    origin,
+    chainId,
+    autoApprove,
+  ) {
```

Comment:
nit - I find it easier to understand these functions inline when the args are passed in as object rather than by position - like in this case where this is used in `ethereum-chain-utils` you can see that the `isAddFlow` boolean is used for the `autoApprove` arg without having toggle between the declaration to understand ... I suppose the docs also help with this. Anyways I know we do it both ways throughout the codebase. Just putting in for my preference 😅

Reply:
https://github.com/MetaMask/metamask-extension/pull/29213/commits/c70f7b7d29ad5ac28aff03fdfe121e16d40a8440